cpp/ql/src/Likely Bugs/Underspecified Functions/MistypedFunctionArguments.ql

```diff
@@ -0,0 +1,109 @@
+/**
+ * @name Call to a function with one or more incompatible arguments
+ * @description A call to a function with at least one argument whose type does
+ * not match the type of the corresponding function parameter.  This may indicate
+ * that the author is not familiar with the function being called.  Passing mistyped
+ * arguments on a stack may lead to unpredictable function behavior.
+ * @kind problem
+ * @problem.severity warning
+ * @precision very-high
+ * @id cpp/mistyped-function-arguments
+ * @tags correctness
+ *       maintainability
+ */
+
+import cpp
+
+pragma[inline]
+int sizeofInt() { result = any(IntType pt).getSize() }
+
+pragma[inline]
+predicate pointerArgTypeMayBeUsed(Type arg, Type parm) {
+  arg = parm
+  or
+  // arithmetic types
+  arg instanceof ArithmeticType and
+  parm instanceof ArithmeticType and
+  arg.getSize() = parm.getSize() and
+  (
+    (
+      arg instanceof IntegralType and
+      parm instanceof IntegralType
+    )
+    or
+    (
+      arg instanceof FloatingPointType and
+      parm instanceof FloatingPointType
+    )
+  )
+  or
+  // pointers to void are ok
+  arg instanceof VoidType
+  or
+  parm instanceof VoidType
+}
+
+pragma[inline]
+predicate argTypeMayBeUsed(Type arg, Type parm) {
+  arg = parm
+  or
+  // float will be promoted to double, and so it should correspond
+  // to the prototype
+  arg instanceof FloatType and parm instanceof DoubleType
+  or
+  // integral types are promoted "up to" (unsigned) int, but not long long.
+  arg instanceof IntegralType and
+  parm instanceof IntegralType and
+  arg.getSize() <= sizeofInt() and
+  parm.getSize() <= sizeofInt()
```

Comment:
Now that you're calling `getFullyConverted` on the argument, you don't need to emulate promotions of the argument in the query, only promotions of the parameter. Parameter promotions happen in K&R-style function definitions only. I don't know if we can tell from the database whether the definition was K&R-style or had a prototype; if not, we'll have to assume that both are possible.

Reply:
Should we even worry about K&R parameter promotions -- at least now?  I'd leave it for now.  We can always tighten our query later if we see this out in the field.

Reply:
Turns out still need to establish <code>IntegralType</code>-ness (though perhaps not the size) of arguments and parameters.  Otherwise, we get
![image](https://user-images.githubusercontent.com/48528981/55513602-f4572100-561a-11e9-8ccb-f26bd1ea0b5a.png)